Sources/Raft/Consensus/Consensus.swift

```diff
@@ -0,0 +1,227 @@
+
+        // TODO Handle errors correctly
+        let result = try! await Task.withGroup(resultType: Bool.self, returning: Bool.self) { group in
+
+            // We should stop election at the moment when we got quorum
+            for peer in peers {
+                // We should stop election at the moment when we got quorum
+                await group.add(operation: {
+                    await peer.requestVote(request).voteGranted
+                })
+            }
+
+            do {
+                var grantedVotes: UInt = 1 // We already votes for ourself
+                while let result = try await group.next() {
+                    grantedVotes += result ? 1 : 0
+                    if grantedVotes >= tallyVotes {
+                        return true
+                    }
+                }
+                return false
+            } catch {
+                // If got any error we lost an election
+                return false
+            }
+        }
+        return result
+    }
+
+    /// Vote request message. Use for both pre and real vote
+    public struct RequestVote {
+
+        public struct Request {
+            /// Vote type `vote` or `PreVote`
+            let type: VoteType
+
+            /// Candidate’s term
+            let term: Term.Id
+
+            /// Candidate requesting vote
+            let candidate: NodeId
+
+            /// Index of candidate’s last log entry
+            let lastLogIndex: UInt64
+
+            /// Term of candidate’s last log entry
+            let lastLogTerm: UInt64
+
+        }
+
+        public struct Response {
+            /// Vote type `vote` or `PreVote`, should be the the same as in request
+            let type: VoteType
+
+            /// Current term of the node, for candidate to update itself
+            let term: Term
+
+            /// True means candidate received vote
+            let voteGranted: Bool
+        }
+    }
+
+    /// Process vote and pre vote requests. Should be invoked by a host
+    ///
+    /// 1. Reply false if term `<` currentTerm (§5.1)
+    /// 2. If votedFor is null or candidateId, and candidate’s log is at least as up-to-date as receiver’s log, grant vote (§5.2, §5.4)
+    func onVoteRequest(_ request: RequestVote.Request) async -> RequestVote.Response {
+        // TODO: On vote response we also should reset election timer
+        let granted: Bool = {
+            if request.type == .preVote {
+                return request.term > term.id
+            } else {
+                return term.canAcceptNewTerm(request.term, from: request.candidate)
+            }
+        }()
+        logger.debug("Vote response \(granted)", metadata: [
+            "vote/self": "\(myself)",
+            "vote/candidate": "\(request.candidate)",
+            "vote/granted": "\(granted)"
+        ])
+        return .init(type: request.type, term: self.term, voteGranted: granted)
+    }
+
+}
+
+// MARK: Entries
+extension Consensus {
```

Comment:
I guess I would call this `Raft` or something else perhaps? Consensus is a bit weird name for it hm

Reply:
(it'd conflict with the module name, but generally we try to prefer nice type names, and less nice module names; so `SwiftRaft` could be a nice module name etc)

Reply:
Before I have used `ConsensusService`, but it also looks wired and naming is hard :) Will update module name to `SwiftRaft` and actor to `Raft`